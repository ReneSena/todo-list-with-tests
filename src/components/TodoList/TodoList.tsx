import React, { FC } from 'react';
import styled from 'styled-components';

export interface ITodoListItem {
    id: string;
    value: string;
}

export interface ITodoList {
    listItems: ITodoListItem[];
}
interface IProps {
    done: boolean;
}

const ListItem = styled.li<IProps>`
    text-decoration: ${(props) => props.done && 'line-through'};
`;

const TodoListItem: FC<ITodoListItem> = ({ id, value }) => {
    const [taskDone, setTaskDone] = React.useState<boolean>(false);

    const handleCheckbox = () => setTaskDone(!taskDone);

    return (
        <>
            <ListItem id={id} done={taskDone}>
                <input type="checkbox" onClick={handleCheckbox} />
                {value}
            </ListItem>
        </>
    );
};

export const TodoList: FC<ITodoList> = ({ listItems }) => {
    return (
        <>
            {!listItems.length && <div>You have no tasks registered.</div>}
            <ul>
                {listItems.map((item) => (
                    <TodoListItem key={item.id} id={item.id} value={item.value} />
                ))}
            </ul>
        </>
    );
};

