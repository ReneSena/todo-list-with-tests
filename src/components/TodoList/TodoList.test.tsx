import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { TodoList } from './TodoList';

const mockTasks = [
    {
        id: '1234',
        value: 'To do coffe',
    },
    {
        id: '2345',
        value: 'Create a blog',
    },
    {
        id: '3456',
        value: 'Drink water',
    },
];

describe('<TodoList />', () => {
    afterAll(() => jest.resetAllMocks());
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render TodoList Component with empty list', () => {
        render(<TodoList listItems={[]} />);

        expect(screen.getByRole('list')).toBeInTheDocument();
        expect(screen.getByText('You have no tasks registered.')).toBeInTheDocument();
    });

    it('should render TodoList Component with tasks', () => {
        render(<TodoList listItems={mockTasks} />);

        expect(screen.getAllByRole('list')).toMatchSnapshot();
        expect(screen.getAllByRole('listitem')).toHaveLength(3);
    });
});

