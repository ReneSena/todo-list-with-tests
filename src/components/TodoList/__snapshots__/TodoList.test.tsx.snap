// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> should render TodoList Component with tasks 1`] = `
Array [
  <ul>
    <li
      class="sc-bczRLJ fThpee"
      id="1234"
    >
      <input
        type="checkbox"
      />
      To do coffe
    </li>
    <li
      class="sc-bczRLJ fThpee"
      id="2345"
    >
      <input
        type="checkbox"
      />
      Create a blog
    </li>
    <li
      class="sc-bczRLJ fThpee"
      id="3456"
    >
      <input
        type="checkbox"
      />
      Drink water
    </li>
  </ul>,
]
`;
